### PR TITLE
port date-parser() from incubator

### PR DESCRIPTION
**Pull request by bazsi**
This series of patches port the the date-parser() from the incubator with a number of changes to make that production grade.

Changes from incubator:
* removed the offset() option, as it can be replaced by template functions
* fixed a bug in clone() that would have prevented using a date-parser() instance from multiple locations in the configuration
* if no timezone information is known, use the local timezone (whereas the incubator used GMT)
* use the same year deduction logic as in syslog-format
* expect the entire input to be parsed, if remainding characters are present, assume it was not in the proper format and fail parsing
* uses a private strptime() implementation ported from NetBSD with some fixups, it _always_ uses the C locale, and properly supports %z/%Z
* refactors to make the code more readable/maintainable in the long term

@vincentbernat, please review at least the functionality changes, but I'd appreciate if you could have a glimpse on the code as well.

Please merge, thanks.

### CODE REVIEW

**Comment by vincentbernat** [comment]
I am fine with all the changes. The `%Z` support is quite restrictive, but it's better than no support at all. Maybe a note in the documentation discouraging its use due to the lack of a comprehensive timezone database?

**Comment by bazsi** [comment]
Yeah, documenting the strptime format and restrictions in details is a good
idea.

**Comment by faxm0dem** [comment]
I have a question about the local tz being used in favour of UTC. Is it still possible to force the zone in the block using `time_zone()`?

**Comment by bazsi** [comment]
Yeah, there is an explicit time-zone() option.

I have a question about the local tz being used in favour of UTC. Is it
still possible to force the zone in the block using time_zone()?

—
Reply to this email directly or view it on GitHub
<https://github.com/balabit/syslog-ng/pull/755#issuecomment-152655702>.

**Comment by faxm0dem** [comment]
Awesome, thanks !

File: modules/date/date-parser.c [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+  /* initialize tm with current date, this fills in dst and other
+   * fields (even non-standard ones) */
+
+  cached_localtime(&now, &tm);
+
+  if (!_parse_timestamp_and_deduce_missing_parts(self, &tm, &tm_zone_offset, input))
+    return FALSE;
+
+  if (!_convert_struct_tm_to_logstamp(self, now, &tm, tm_zone_offset, target))
```
Comment (ihrwein) [inline modules/date/date-parser.c]: it's just a nitpick, but a simple `return _convert_struct_tm_to_logstamp()` from here?

**Comment by ihrwein** [comment]
:+1:

File: modules/date/date-parser.c (hunk 1 above)
Comment (bazsi) [inline modules/date/date-parser.c]: I could find both pro and con arguments for this. Let's see if there are other comments, if I do need to touch the code I'll address this az well.

Comment (ihrwein) [inline modules/date/date-parser.c]: Ok.

**Pull request merged**